packages/node_modules/node-red/red.js

```diff
@@ -142,208 +142,278 @@ if (process.env.NODE_RED_ENABLE_PROJECTS) {
+    
+        // Refresh https settings at intervals for NodeJs version 11 and above
+        if (settings.httpsRefreshInterval) {
+            if (typeof settings.https === "function") {        
+                if (server.setSecureContext) {
+                    console.log("Refreshing https settings every " + parseInt(settings.httpsRefreshInterval) + " seconds.");
+                    setInterval(function () {
+                        try {
+                            // Get the result of the function, because createServer doesn't accept functions as input
+                            var refreshedHttps = settings.https();
+                            
+                            var httpsPromise;
+                            if (refreshedHttps.hasOwnProperty('then') && typeof refreshedHttps.then === 'function') {
+                                // A promise was returned
+                                httpsPromise = refreshedHttps;
+                            } else {
+                                // An object was returned - wrap in a promise
+                                httpsPromise = Promise.resolve(refreshedHttps);
+                            }
 
-if (settings.httpAdminRoot !== false) {
-    settings.httpAdminRoot = formatRoot(settings.httpAdminRoot || settings.httpRoot || "/");
-    settings.httpAdminAuth = settings.httpAdminAuth || settings.httpAuth;
-} else {
-    settings.disableEditor = true;
-}
+                            httpsPromise.then(function(refreshedHttps) {
+                                // Use the refreshed https settings
+                                if (!refreshedHttps.key || !refreshedHttps.cert) {
+                                    console.log("Cannot refresh the https settings when the https property function doesn't return a 'key' and 'cert'.");
+                                    return;
+                                }
+                                    
+                                // Only update the credentials in the server when key or cert has changed
+                                if(!server.key || !server.cert || !server.key.equals(refreshedHttps.key) || !server.cert.equals(refreshedHttps.cert)) {
+                                    server.setSecureContext(refreshedHttps);
+                                    console.log("The https settings have been refreshed.");
+                                }
+                            }).catch(function(err) {
+                                console.log("Failed to apply the refreshed https settings: " + err);
+                            });
+                        } catch(err) {
+                            console.log("Failed to get the refreshed https settings: " + err);
+                        }
+                    }, parseInt(settings.httpsRefreshInterval)*1000);
+                } else {
+                    console.log("Cannot refresh the https settings automatically, because NodeJs version 11 or above is required.");
```

Comment:
See my other comment about deferring the logging of this error until later in the file - that will allow the message to be translated etc.

Reply:
Done